ui-modules/blueprint-composer/app/components/util/d3-blueprint.js

```diff
@@ -771,50 +774,37 @@ export function D3Blueprint(container, options) {
         return node;
     }
 
+    /**
+     * Draws new relationships as arcs and labels. Updates existing ones.
+     */
     function drawRelationships() {
+
         showRelationships();
 
-        let relationData = _relationGroup.selectAll('.relation')
-            .data(_d3DataHolder.visible.relationships, (d)=>(d.source._id + '_related_to_' + d.target._id));
+        const getPathId = (d) => (d.source._id + '-' + d.target._id);
+        const getRelationId = (d) => (getPathId(d) + '-' + d.pathSelector);
+
+        // ====== RELATIONSHIP ARCS ===========
 
-        let relationDataEntered = relationData.enter();
+        let arcsData = Object.values(_d3DataHolder.visible.relationships.reduce((accumulator, d) => {
+            accumulator[getRelationId(d)] = {
+                source: d.source,
+                target: d.target,
+                pathSelector: d.pathSelector
+            };
+            return accumulator;
+        }, {}));
 
-        // Draw the relationship path
-        relationDataEntered.insert('path')
+        let relationArcs = _relationArcs.selectAll('.relation').data(arcsData, (d) => getRelationId(d));
+
+        relationArcs.enter().insert('path')
```

Comment:
Are we inserting a list of `<path>` elements to the DOM here?

If so, it would probably be more performant (and angular-like) to have the template html iterate something like `$scope.relationships` to generate them.

A `$watch` statement could keep that scope variable updated whenever any of the dependant properties for the `drawRelationships` logic changes.

Reply:
@jathanasiou, this is not angular, this is D3, not sure if it applies.

Reply:
Haven't touched D3 in a long time so not sure about how it interacts with the DOM. If D3 itself is calling `drawRelationships` and dealing with DOM updates, yeah it's probably not a good idea to bring angular templates into this.

Reply:
It's just we will need to compile angular components on a fly to inject them here, which is not a good idea.